url433bec1e56_sarobe_DifficultyInvestigation_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/gnuplot/Grapher.java

```diff
@@ -0,0 +1,107 @@
+package gnuplot;
+
+import common.Params;
+import common.utilities.JEasyFrame;
+
+import javax.swing.*;
+import java.io.*;
+import java.util.Arrays;
+
+/**
+ * Created by Samuel Roberts, 2012
+ */
+public class Grapher {
+
+    public static void main(String[] args) throws Exception {
+
+        final int startingIndex = 56;
+        final int graphs = 1;
+
+        for(int i=startingIndex; i<startingIndex + graphs; i++) {
+//            transformData(i, 0);
+            drawGraph(i);
+        }
+        new JEasyFrame(new JPanel(), "Keep Graph Open!");
+    }
+
+    public static void drawGraph(int runIndex) {
+        try {
+            GnuPlot gp = new GnuPlot();
+            gp.setAxes("Generation", "Std Devs of Improvement");
+            gp.plot(Params.dataDirectory + "run-" + runIndex + "/gnuplot-best.txt", "Best of Generation");
+//            gp.replot(Params.dataDirectory + "run-" + runIndex + "/gnuplot-mean.txt", "Mean of Generation");
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    public static void writeGenData(int runIndex, int genIndex, double bestResult, double meanResult, double[] bestSolution) throws IOException {
+
+        // Use to write gnuplot compatible data
+
+        // make directory to write to
+        String directoryName = Params.dataDirectory + "run-" + runIndex;
+        new File(directoryName).mkdir();
+
+        // write/append data to graphing files
+        boolean append = true;
+        if(genIndex == 1) {
+            append = false;
+        }
+        // write the best
+        PrintWriter pw = new PrintWriter(new FileWriter(directoryName + "/gnuplot-best.txt", append));
+        pw.format("%f \t %f \n", (double)genIndex, bestResult);
+        pw.close();
+
+        // write the mean
+        pw = new PrintWriter(new FileWriter(directoryName + "/gnuplot-mean.txt", append));
+        pw.format("%f \t %f \n", (double)genIndex, meanResult);
+        pw.close();
+
+        // write the best ship
+        pw = new PrintWriter(new FileWriter(directoryName + "/best-ship.txt", append));
+        pw.print(genIndex + "\t" + bestResult + "\t" + Arrays.toString(bestSolution) + "\n");
+        pw.close();
+    }
+
+    public static void transformData(int runIndex, double constant) throws Exception {
+        String directoryName = Params.dataDirectory + "run-" + runIndex;
+
+        // rewrite the best
+        BufferedReader reader = new BufferedReader(new FileReader(directoryName + "/gnuplot-best.txt"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced at lines 71, 74. The resource is closed at line 80. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).